ProcessLib/ThermoHydroMechanics/ThermoHydroMechanicsFEM.h

```diff
@@ -0,0 +1,894 @@
+                _process_data.getFluidHeatCapacity(
+                        p_int_pt, T_int_pt);
+            double const C_s =
+                _process_data.solid_specific_heat_capacity(t, x_position)[0];
+            double const T0 =
+                _process_data.reference_temperature(t, x_position)[0];
+            auto const alpha = _process_data.biot_coefficient(t, x_position)[0];
+            auto rho_sr = _process_data.solid_density(t, x_position)[0];
+            auto rho_f = _process_data.getFluidDensity(
+                        p_int_pt, T_int_pt);
+            auto const porosity = _process_data.porosity(t, x_position)[0];
+            auto const& b = _process_data.specific_body_force;
+
+            // calculate linear thermal strain
+            // assume isotropic thermal expansion
+
+            double delta_T(T_int_pt - T0);
+            double const thermal_strain = alpha_s * delta_T;
+
+            //double rho_f = rho_fr * (1 - beta_f * delta_T);
+            double rho_s = rho_sr * (1 - 3 * thermal_strain);
+
+            auto velocity = (-K_over_mu * dNdx_p * p).eval();
+            velocity += K_over_mu * rho_f * b;
+
+            // mapping matrix (m)
+            auto const& identity2 = MaterialLib::SolidModels::Invariants<
+                KelvinVectorDimensions<DisplacementDim>::value>::identity2;
+
+            //
+            // displacement equation, displacement part (K_uu)
+            //
+            eps.noalias() = B * u;
+
+            auto C = _ip_data[ip].updateConstitutiveRelation(t, x_position, dt,
+                                                    thermal_strain);
+
+            local_Jac
+                .template block<displacement_size, displacement_size>(
+                    displacement_index, displacement_index)
+                .noalias() += B.transpose() * C * B * w;
+
+            double const rho = rho_s * (1 - porosity) + porosity * rho_f;
+            local_rhs.template segment<displacement_size>(displacement_index)
+                .noalias() -=
+                (B.transpose() * sigma_eff - N_u_op.transpose() * rho * b) * w;
+
+            //
+            // displacement equation, pressure part (K_up)
+            //
+            Kup.noalias() += B.transpose() * alpha * identity2 * N_p * w;
+
+            //
+            // pressure equation, pressure part (K_pp and M_pp).
+            //
+            laplace_p.noalias() += dNdx_p.transpose() * K_over_mu * dNdx_p * w;
+
+            storage_p.noalias() += N_p.transpose() * S * N_p * w;
+            //
+            // fp
+            //
+            local_rhs.template segment<pressure_size>(pressure_index)
+                .noalias() += dNdx_p.transpose() * rho_f * K_over_mu * b * w;
+            //
+            // pressure equation, temperature part (M_pT)
+            //
+            auto beta = porosity * beta_f + (1-porosity)* 3 * alpha_s;
+            storage_T.noalias() += N_p.transpose() * beta * N_p * w;
```

Comment:
Shouldn't it be \int_\Omega N_T * beta * N_T d\Omega?

Reply:
corrected